vic/drivers/shared_all/src/calc_root_fraction.c

```diff
@@ -41,128 +41,126 @@ calc_root_fractions(veg_con_struct  *veg_con,
+
+                // Reset running total for next layer
+                sum_fract = 0;
+
+            }
+
+            // Decide whether to increment layer or zone
+            if (Lsum < Zsum) {
+                // zone extends beyond layer, so increment layer
                 layer++;
                 if (layer < options.Nlayer) {
-                    Lstep = soil_con->depth[layer];
-                    Lsum += Lstep;
+                    Lsumprev = Lsum;
+                    Lsum += soil_con->depth[layer];
                 }
-                else if (layer == options.Nlayer && zone < options.ROOT_ZONES) {
-                    Zstep = (double) veg_con[veg].zone_depth[zone];
-                    Lstep = Zsum + Zstep - Lsum;
-                    if (zone < options.ROOT_ZONES - 1) {
-                        for (i = zone + 1; i < options.ROOT_ZONES; i++) {
-                            Lstep += veg_con[veg].zone_depth[i];
-                        }
-                    }
-                    Lsum += Lstep;
+                else {
+                    Lsum = Zsum;
                 }
             }
-            else if (Zsum + Zstep > Lsum) {
+            else if (Lsum > Zsum) {
+                // layer extends beyond zone, so increment zone
                 zone++;
-                if (layer < options.Nlayer) {
-                    veg_con[veg].root[layer] = sum_fract;
-                    sum_fract = 0.;
+                if (zone < options.ROOT_ZONES) {
+                    Zsum += veg_con[veg].zone_depth[zone];
+                }
+                else {
+                    Zsum = Lsum;
                 }
+            }
+            else {
+                // layer and zone end at the same depth, so increment both
                 layer++;
                 if (layer < options.Nlayer) {
-                    Lstep = soil_con->depth[layer];
-                    Lsum += Lstep;
+                    Lsumprev = Lsum;
+                    Lsum += soil_con->depth[layer];
                 }
-                else if (layer == options.Nlayer) {
-                    Lstep = Zsum + Zstep - Lsum;
-                    if (zone < options.ROOT_ZONES - 1) {
-                        for (i = zone + 1; i < options.ROOT_ZONES; i++) {
-                            Lstep += veg_con[veg].zone_depth[i];
-                        }
-                    }
-                    Lsum += Lstep;
+                zone++;
+                if (zone < options.ROOT_ZONES) {
+                    Zsum += veg_con[veg].zone_depth[zone];
                 }
             }
-        }
 
-        if (sum_fract > 0 && layer >= options.Nlayer) {
-            veg_con[veg].root[options.Nlayer - 1] += sum_fract;
-        }
-        else if (sum_fract > 0) {
-            veg_con[veg].root[layer] += sum_fract;
         }
 
+        // Final check on root fractions. If they don't sum to 1, throw error
+        // Otherwise, rescale by sum to eliminate small rounding errors
```

Comment:
I think it would be a good idea to also add a check to see whether the sum of `veg_con[veg].root[layer]` is close to one before rescaling (and throw an error if not), to prevent input/algorithm error

Reply:
OK, will do.